File: src/Infrastructure/Repositories/ProjectRepository.cs
using Application.Projects;

using Domain.Entities;

using Infrastructure.Data;

using Microsoft.EntityFrameworkCore;

namespace Infrastructure.Repositories;

public class ProjectRepository(AppDbContext db) : IProjectRepository
{
    public async Task<IEnumerable<ProjectDto>> GetAllAsync(CancellationToken cancellationToken = default)
        => await db.Projects
            .Select(p => new ProjectDto(
                p.Id, p.Name, p.Description, p.CreatedAt, p.UpdatedAt,
                p.TestSuites.Count(s => !s.IsDeleted),
                p.TestRuns.Count(r => !r.IsDeleted)))
            .ToListAsync(cancellationToken);

    public async Task<ProjectDto?> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
        => await db.Projects
            .Where(p => p.Id == id)
            .Select(p => new ProjectDto(
                p.Id, p.Name, p.Description, p.CreatedAt, p.UpdatedAt,
                p.TestSuites.Count(s => !s.IsDeleted),
                p.TestRuns.Count(r => !r.IsDeleted)))
            .FirstOrDefaultAsync(cancellationToken);

    public Task<bool> ExistsAsync(Guid id, CancellationToken cancellationToken = default)
        => db.Projects.AnyAsync(p => p.Id == id, cancellationToken);

    public async Task<ProjectDto> AddAsync(Project project, CancellationToken cancellationToken = default)
    {
        db.Projects.Add(project);
        await db.SaveChangesAsync(cancellationToken);
        return new ProjectDto(project.Id, project.Name, project.Description, project.CreatedAt, project.UpdatedAt);
    }

    public async Task<bool> UpdateAsync(Guid id, Action<Project> mutate, CancellationToken cancellationToken = default)
    {
        var project = await db.Projects.FirstOrDefaultAsync(p => p.Id == id, cancellationToken);
        if (project is null) return false;

        mutate(project);
        await db.SaveChangesAsync(cancellationToken);
        return true;
    }

    public async Task<bool> DeleteAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var project = await db.Projects.FirstOrDefaultAsync(p => p.Id == id, cancellationToken);
        if (project is null) return false;

        project.SoftDelete();
        await db.SaveChangesAsync(cancellationToken);
        return true;
    }
}

File: apps/api/src/infrastructure/repositories/project.repository.ts
import { Paginated, PaginationParams } from "@testcraft/types";

import {
  CreateProject,
  IProjectRepository,
  UpdateProject,
} from "@/application/projects/project.repository";
import { resolvePagination } from "@/domain/pagination";
import { Project } from "@/domain/project";
import { PrismaClient } from "@/generated/prisma/client";
import { isNotFound } from "@/infrastructure/database/prisma.errors";

const projectSelect = {
  id: true,
  userId: true,
  name: true,
  description: true,
  createdAt: true,
  updatedAt: true,
  _count: {
    select: {
      testSuites: { where: { isDeleted: false } },
      testRuns: { where: { isDeleted: false } },
    },
  },
} as const;

const toDto = (project: {
  id: string;
  userId: string;
  name: string;
  description: string | null;
  createdAt: Date;
  updatedAt: Date;
  _count: { testSuites: number; testRuns: number };
}): Project => ({
  id: project.id,
  userId: project.userId,
  name: project.name,
  description: project.description,
  createdAt: project.createdAt,
  updatedAt: project.updatedAt,
  suiteCount: project._count.testSuites,
  runCount: project._count.testRuns,
});

export class ProjectRepository implements IProjectRepository {
  constructor(private readonly prisma: PrismaClient) {}

  async getAll(
    userId: string,
    search?: string,
    pagination?: PaginationParams,
  ): Promise<Paginated<Project>> {
    const { page, pageSize, skip, take } = resolvePagination(pagination);
    const where = {
      userId,
      isDeleted: false,
      ...(search
        ? { name: { contains: search, mode: "insensitive" as const } }
        : {}),
    };
    const [rows, total] = await this.prisma.$transaction([
      this.prisma.project.findMany({
        where,
        select: projectSelect,
        orderBy: { createdAt: "desc" },
        skip,
        take,
      }),
      this.prisma.project.count({ where }),
    ]);
    return { items: rows.map(toDto), total, page, pageSize };
  }

  async getById(userId: string, id: string): Promise<Project | null> {
    const project = await this.prisma.project.findFirst({
      where: { id, userId, isDeleted: false },
      select: projectSelect,
    });
    return project ? toDto(project) : null;
  }

  async create(userId: string, input: CreateProject): Promise<Project> {
    const project = await this.prisma.project.create({
      data: {
        userId,
        name: input.name,
        description: input.description ?? null,
      },
      select: projectSelect,
    });
    return toDto(project);
  }

  async update(
    userId: string,
    id: string,
    input: UpdateProject,
  ): Promise<Project | null> {
    try {
      const project = await this.prisma.project.update({
        where: { id, userId, isDeleted: false },
        data: { name: input.name, description: input.description ?? null },
        select: projectSelect,
      });
      return toDto(project);
    } catch (err) {
      if (isNotFound(err)) return null;
      throw err;
    }
  }

  async delete(userId: string, id: string): Promise<boolean> {
    try {
      await this.prisma.project.update({
        where: { id, userId, isDeleted: false },
        data: { isDeleted: true, deletedAt: new Date() },
      });
      return true;
    } catch (err) {
      if (isNotFound(err)) return false;
      throw err;
    }
  }
}
